backend_new/src/services/application-csv-export.service.ts

```diff
@@ -0,0 +1,596 @@
+                // return preference/program as value for the format function to accept
+                if (multiselectQuestionValue) {
+                  return acc;
+                }
+
+                if (parsePreference) {
+                  // curr should equal the preference id we're pulling from
+                  if (!preferences) {
+                    preferences = JSON.parse(app.preferences as string);
+                  }
+                  parsePreference = false;
+                  // there aren't typically many preferences, but if there, then a object map should be created and used
+                  const preference = preferences.find(
+                    (preference) => preference.multiselectQuestionId === curr,
+                  );
+                  multiselectQuestionValue = true;
+                  return preference;
+                }
+
+                if (parsePrograms) {
+                  // curr should equal the program id we're pulling from
+                  if (!programs) {
+                    programs = JSON.parse(app.programs as string);
+                  }
+                  parsePrograms = false;
+                  const program = programs.find(
+                    (preference) => preference.multiselectQuestionId === curr,
+                  );
+                  multiselectQuestionValue = true;
+                  return program;
+                }
+
+                // sets parsePreference to true, for the next iteration
+                if (curr === 'preferences') {
+                  parsePreference = true;
+                }
+                // sets parsePrograms to true, for the next iteration
+                if (curr === 'programs') {
+                  parsePrograms = true;
+                }
+
+                // handles working with arrays, e.g. householdMember.0.firstName
+                if (!isNaN(Number(curr))) {
+                  const index = Number(curr);
+                  return acc[index];
+                }
+
+                if (acc === null || acc === undefined) {
+                  return '';
+                }
+                return acc[curr];
+              }, app);
+              value = value === undefined ? '' : value === null ? '' : value;
+              if (header.format) {
+                value = header.format(value);
+              }
+
+              row += value;
+              if (index < csvHeaders.length - 1) {
+                row += ',';
+              }
+            });
+
+            try {
+              writableStream.write(row + '\n');
+            } catch (e) {
+              console.log('writeStream write error = ', e);
+              writableStream.once('drain', () => {
+                console.log('drain buffer');
+                writableStream.write(row + '\n');
+              });
+            }
+          });
+
+          writableStream.end();
+        });
+    });
+  }
+
+  private async maxHouseholdMembers(): Promise<number> {
```

Comment:
I think making this a public function as well, makes unit testing functions like this a lot easier

Reply:
made public